lib/connection.js

```diff
@@ -13,220 +12,218 @@ const UTP_ERRORS = [
+      this._onend.bind(this),
+      this._onerror.bind(this),
+      this._onclose.bind(this),
+      this._onconnect.bind(this),
+      this._realloc.bind(this)
+    )
+
+    unordered.add(utp.connections, this)
+    if (utp.maxConnections && utp.connections.length >= utp.maxConnections) {
+      utp.firewall(true)
+    }
+  }
 
-Connection.prototype.address = function () {
-  if (this.destroyed) return null
-  return this._utp.address()
-}
+  setTimeout (ms, ontimeout) {
+    if (ontimeout) this.once('timeout', ontimeout)
+    if (this._timeout) this._timeout.destroy()
+    this._timeout = timeout(ms, this._ontimeout, this)
+  }
 
-Connection.prototype._read = function () {
-  // TODO: backpressure
-}
+  _ontimeout () {
+    this.emit('timeout')
+  }
 
-Connection.prototype._write = function (data, enc, cb) {
-  if (this.destroyed) return
+  setInteractive (interactive) {
+    this.setPacketSize(interactive ? 0 : 65536)
+  }
 
-  if (!this._connected || !binding.utp_napi_connection_write(this._handle, data)) {
-    this._callback = cb
-    this._writing = new Array(1)
-    this._writing[0] = data
-    return
+  setContentSize (size) {
+    this._view[0] = size < 65536 ? (size >= 0 ? size : 0) : 65536
+    this._contentSize = size
   }
 
-  cb(null)
-}
+  setPacketSize (size) {
+    if (size > 65536) size = 65536
+    this._view[0] = size
+    this._contentSize = 0
+  }
 
-Connection.prototype._writev = function (datas, cb) {
-  if (this.destroyed) return
+  address () {
+    if (this.destroyed) return null
+    return this._utp.address()
+  }
 
-  const bufs = new Array(datas.length)
-  for (var i = 0; i < datas.length; i++) bufs[i] = datas[i].chunk
+  _read (cb) {
+    // TODO: backpressure
+    // cb(null)
+  }
 
-  if (bufs.length > 256) return this._write(Buffer.concat(bufs), null, cb)
+  _write (data, cb) {
+    if (typeof data === 'string') data = Buffer.from(data)
+    if (!this._connected || !binding.utp_napi_connection_write(this._handle, data)) {
+      this._callback = cb
+      this._writing = new Array(1)
+      this._writing[0] = data
+      return
+    }
 
-  if (!binding.utp_napi_connection_writev(this._handle, bufs)) {
-    this._callback = cb
-    this._writing = bufs
-    return
+    cb(null)
   }
 
-  cb(null)
-}
+  _writev (datas, cb) {
+    const bufs = new Array(datas.length)
+    for (var i = 0; i < datas.length; i++) {
+      const data = datas[i].chunk
+      bufs[i] = typeof data === 'string' ? Buffer.from(data) : data
+    }
 
-Connection.prototype._realloc = function () {
-  this._buffer = Buffer.allocUnsafe(this._buffer.length)
-  this._offset = 0
-  return this._buffer
-}
+    if (bufs.length > 256) return this._write(Buffer.concat(bufs), cb)
 
-Connection.prototype._onread = function (size) {
-  if (!this._connected) this._onconnect() // makes the server wait for reads before writes
-  if (this._timeout) this._timeout.refresh()
+    if (!binding.utp_napi_connection_writev(this._handle, bufs)) {
```

Comment:
this needs the !this._connected check also cause streamx will eagerly use this hook